### PR TITLE
Stable radical optimization merge

### PR DESCRIPTION
**Pull request by jlaw9**
Attempt to port over the stable radical optimization problem. Haven't done any testing yet. I left "TODO"s where I had questions.

### CODE REVIEW

File: stable_radical_optimization/stable_radical_opt.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+            return MoleculeState(rdkit.Chem.MolFromSmiles('C'), self._config)
+
+        def get_reward(self, state: MoleculeState) -> Tuple[float, dict]:
+            # TODO what does this do? Is it needed?
+            ## Node is outside the domain of validity
+            #elif ((node.policy_inputs['atom'] == 1).any() | 
+            #    (node.policy_inputs['bond'] == 1).any()):
+            #    return config.min_reward
+
+            # TODO what data should be stored here?
```
Comment (pstjohn) [inline stable_radical_optimization/stable_radical_opt.py:73]: see below

File: stable_radical_optimization/stable_radical_opt.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+            # This is a bit of a placeholder; but the range for spin is about 1/50th that
+            # of buried volume.
+            reward = ((1 - max_spin) * 50 + spin_buried_vol 
+                    + 25 / (1 + np.exp(-(bde_diff - 10)))
+                    + 100 * (
+                self.windowed_loss(electron_affinity, ea_range) +     
+                self.windowed_loss(ionization_energy, ie_range) + 
+                self.windowed_loss(v_diff, v_range) +         
+                self.windowed_loss(bde, bde_range)) / 4)
```
Comment (pstjohn) [inline stable_radical_optimization/stable_radical_opt.py:119]: we'd probably want to store all these numbers, `max_spin`, etc., that come out of the predictive models

File: stable_radical_optimization/stable_radical_opt.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    redox_model = tf.keras.models.load_model(redox_model, compile=False)
+    bde_model = tf.keras.models.load_model(bde_model, compile=False)
+
+    # TODO what values should we set for these options?
+    #:param max_atoms: Maximum number of heavy atoms
+    #:param min_atoms: minimum number of heavy atoms
+    #:param atom_additions: potential atom types to consider. Defaults to ('C', 'H', 'O')
+    #:param stereoisomers: whether to consider stereoisomers different molecules
+    #:param sa_score_threshold: If set, don't construct molecules greater than a given sa_score.
+    #:param tryEmbedding: Try to get a 3D embedding of the molecule, and if this fails, remote it.
```
Comment (pstjohn) [inline stable_radical_optimization/stable_radical_opt.py:171]: i think this is OK, sa_score_threshold should be 3.5

File: stable_radical_optimization/stable_radical_opt.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+    problem = StableRadOptProblem(
+        engine,
+        config,
+        stability_model,
+        redox_model,
+        bde_model,
+        run_id=run_id,
+        reward_class=reward_factory,
+        # TODO what values should we set for these options?
```
Comment (pstjohn) [inline stable_radical_optimization/stable_radical_opt.py:214]: Let's start with the defaults from the last run, which would be here
https://github.com/NREL/rlmolecule/blob/augmenting_reward/alphazero/config.py#L25-L27

File: stable_radical_optimization/stable_radical_opt.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+#sys.path.append('/projects/rlmolecule/pstjohn/models/20201031_bde/')
+#from preprocess_inputs import preprocessor as bde_preprocessor
+#bde_preprocessor.from_json('/projects/rlmolecule/pstjohn/models/20201031_bde/preprocessor.json')
+# TODO would we need a different preprocessor.json file here
+# than we have in rlmolecule.molecule.policy.preprocessor.json?
+bde_preprocessor = preprocessor.load_preprocessor()
```
Comment (pstjohn) [inline stable_radical_optimization/stable_radical_opt.py:28]: Yeah, so the BDE model was trained on a different set of data and we need to keep the preprocessor consistent, so we'd want to make sure we can load that bde-specific preprocessor from the bde model directory.

File: stable_radical_optimization/stable_radical_opt.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+        def get_initial_state(self) -> MoleculeState:
+            return MoleculeState(rdkit.Chem.MolFromSmiles('C'), self._config)
+
+        def get_reward(self, state: MoleculeState) -> Tuple[float, dict]:
+            # TODO what does this do? Is it needed?
+            ## Node is outside the domain of validity
+            #elif ((node.policy_inputs['atom'] == 1).any() | 
+            #    (node.policy_inputs['bond'] == 1).any()):
+            #    return config.min_reward
```
Comment (pstjohn) [inline stable_radical_optimization/stable_radical_opt.py:71]: The preprocessor will put a '1' in an atom or bond slot if it doesn't recognize it. (0 is the padding token, and stored atoms start with '2'). So if we see that, we really shouldn't be making a prediction, since it's likely a bogus molecular structure anyways.

**Review by pstjohn — COMMENTED** [review]
Thanks for doing this!! I put in some comments where you left TODO questions, let me know if I missed any

**Comment by jlaw9** [comment]
Thanks @pstjohn, I'll make those changes. 

I added a new "StableRadicalMoleculeState" to add the radical computation when getting the next actions. 

Everything seems to work well until the code runs into a molecule it has already seen before. For some reason, it tries to compute the reward again but fails because the molecule doesn't have an [atom with a radical electron](https://github.com/jlaw9/rlmolecule/blob/jlaw/augmenting_reward_merge/stable_radical_optimization/stable_radical_opt.py#L148-L159). I'm guessing I either implemented something incorrectly in the [next actions function](https://github.com/jlaw9/rlmolecule/blob/jlaw/augmenting_reward_merge/stable_radical_optimization/stable_radical_molecule_state.py#L28), the forced_terminal value isn't getting set correctly (since that's what's used to determine [when to compute the reward](https://github.com/jlaw9/rlmolecule/blob/jlaw/augmenting_reward_merge/stable_radical_optimization/stable_radical_opt.py#L80)), or the state isn't being retrieved even though it should've already been stored. 

Oh, could it be that when the molecule is first seen, it stores the state with `forced_terminal` set to true, and then next time the molecule is seen, `forced_terminal` is false so it's not matching what's in the database? 

Any ideas of why its failing or suggestions of how to fix this?

**Comment by jlaw9** [comment]
Huh, well after some testing, it appears that if I change the run id or clear the tables, then there are no errors while running the games. However, if I start a new rollout process, then it fails during the first game

**Comment by jlaw9** [comment]
I tried merging your [most recent branch](https://github.com/NREL/rlmolecule/pull/53), and no more errors!

File: stable_radical_optimization/stable_radical_opt.py (hunk 4 above)
Comment (jlaw9) [reply stable_radical_optimization/stable_radical_opt.py]: I think I got everything

**Comment by pstjohn** [comment]
This is great!! Would you be able to set up an eagle submission script similar to the one we used in augmenting_reward? Have you tried running this on ~10 or so eagle nodes and looked at the molecules that get built?

**Comment by jlaw9** [comment]
I tried submitting a job this morning, but it's still pending. Should run tonight. I'll let you know how it goes!
